src/xrpld/app/tx/detail/OnChainCredentials.cpp

```diff
@@ -0,0 +1,454 @@
+        if (!ctx.view.exists(keylet::account(issuer)))
+            return tecNO_TARGET;
+        if (ctx.view.exists(keylet::credential(acc, issuer, credType)))
+            return tecDUPLICATE;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialCreate::doApply()
+{
+    std::uint32_t const closeTime =
+        ctx_.view().info().parentCloseTime.time_since_epoch().count();
+
+    auto const sleOwner = view().peek(keylet::account(account_));
+    if (!sleOwner)
+        return tefINTERNAL;
+
+    // Check reserve availability for new object creation
+    {
+        STAmount const reserve{view().fees().accountReserve(
+            sleOwner->getFieldU32(sfOwnerCount) + 1)};
+
+        if (mPriorBalance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    auto const& txCredType(ctx_.tx[sfCredentialType]);
+    Blob const credType{txCredType.begin(), txCredType.end()};
+
+    AccountID const subj = ctx_.tx.isFieldPresent(sfSubject)
+        ? AccountID(ctx_.tx[sfSubject])
+        : account_;
+    bool const issPresent = ctx_.tx.isFieldPresent(sfIssuer);
+    AccountID const issuer =
+        issPresent ? AccountID(ctx_.tx[sfIssuer]) : account_;
+
+    Keylet const kCred = keylet::credential(subj, issuer, credType);
+
+    auto const sleCred = std::make_shared<SLE>(kCred);
+    sleCred->setAccountID(sfSubject, subj);
+    sleCred->setAccountID(sfIssuer, issuer);
+    sleCred->setFieldVL(sfCredentialType, credType);
+
+    if (ctx_.tx.isFieldPresent(sfSignature))
+        sleCred->setFieldVL(sfSignature, ctx_.tx.getFieldVL(sfSignature));
+
+    auto const optExp = ctx_.tx[~sfExpiration];
+    if (optExp)
+    {
+        if (closeTime > *optExp)
+        {
+            JLOG(j_.trace()) << "Malformed transaction: "
+                                "Expiration time is in the past.";
+            return tecEXPIRED;
+        }
+
+        sleCred->setFieldU32(sfExpiration, ctx_.tx.getFieldU32(sfExpiration));
+    }
+
+    if (ctx_.tx.isFieldPresent(sfURI))
+        sleCred->setFieldVL(sfURI, ctx_.tx.getFieldVL(sfURI));
+
+    if (issPresent)
+        sleCred->setFieldU32(sfFlags, lsfAccepted);
+
+    view().insert(sleCred);
+    auto const page = view().dirInsert(
+        keylet::ownerDir(account_), kCred, describeOwnerDir(account_));
+
+    JLOG(j_.trace()) << "Adding Credential to owner directory "
+                     << to_string(kCred.key) << ": "
+                     << (page ? "success" : "failure");
+
+    if (!page)
+        return tecDIR_FULL;
+
+    sleCred->setFieldU64(sfOwnerNode, *page);
+
+    adjustOwnerCount(view(), sleOwner, 1, j_);
+    view().update(sleOwner);
+
+    return tesSUCCESS;
+}
+
+// ------- DELETE --------------------------
+NotTEC
+CredentialDelete::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureOnChainCredentials))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
```

Comment:
I think it's just this

>         if (!ctx.tx.isFieldPresent(sfSubject) && !ctx.tx.isFieldPresent(sfIssuer))
>         {
>            // Neither field is present, the transaction is malformed.
>            JLOG(ctx.j.trace()) << "Malformed transaction: "
>                           "Invalid Subject and Issuer fields combination.";
>            return temMALFORMED;
>         }

And don't need to check for the valid accounts.

Reply:
There are checks for right combinations of optional fields and their values. Which are not covered by standard transaction checks, as tests show.